### PR TITLE
Network: Cleanup and refactoring PhysicsEngine

### PR DESCRIPTION
**Pull request by wimrijnders**
Some work done in preparation for #3468, while examining the code for better understanding. The logic remains unchanged.

Note that this does not solve #3468, it's just a preliminary cleanup.

### CODE REVIEW

File: lib/network/modules/PhysicsEngine.js [hunk 1]
```diff
@@ -121,14 +121,13 @@ class PhysicsEngine {
       this.stopSimulation(false);
       this.body.emitter.off();
     });
-    // this event will trigger a rebuilding of the cache everything. Used when nodes or edges have been added or removed.
     this.body.emitter.on("_dataChanged", () => {
-      // update shortcut lists
+      // Nodes an/or edges have been added or removed, update shortcut lists.
```
Comment (yotamberk) [inline lib/network/modules/PhysicsEngine.js]: `an/or` => and/or

Comment (wimrijnders) [reply lib/network/modules/PhysicsEngine.js]: OK. Can you hold off merging until @mbroad has reviewed, please?

File: lib/network/modules/components/physics/BarnesHutSolver.js [hunk 2]
```diff
@@ -62,17 +64,27 @@ class BarnesHutSolver {
...
     }
   }
 
 
   /**
+   * @param {Object} parentBranch
+   * @param {Node} node
+   * @private
+   */
+  _getForceContributions(parentBranch, node) {
```
Comment (mbroad) [inline lib/network/modules/components/physics/BarnesHutSolver.js]: A function beginning with `_get` that has no return value...
Now I know this isn't your fault, as you're being consistent with `_getForceContribution`, but this seems wrong.

**Comment by wimrijnders** [comment]
Yeah, tough one to unit test on, because everything should remain as is. I consider the examples in the unit tests running to be a good indication of fitness - actually all tests that run physics.

I did a visual check by running `worldldCupPerformance.html` with `develop` and this PR side by side.

OK, sooooo....think about a possible unit test here. The premise is `physics unchanged`: so how about setting a given seed and checking after a fixed number of iterations if the output is identical to previous?

The drawback I see to that idea is that it's most likely a single-shot test. If *anything* changes in physics, you can throw the test away.

**Comment by mbroad** [comment]
I just meant testing the inputs, outputs, and side effects of individual functions in isolation.

File: lib/network/modules/PhysicsEngine.js (hunk 1 above)
Comment (yotamberk) [reply lib/network/modules/PhysicsEngine.js]: Of course, that's exactly why I waited

**Comment by wimrijnders** [comment]
Actually, I think a full test is easier to do at this point. This would be an indication that the whole physics thing works as before.

File: lib/network/modules/PhysicsEngine.js [hunk 3]
```diff
@@ -121,14 +121,13 @@ class PhysicsEngine {
...
       this.updatePhysicsData();
     });
 
     // debug: show forces
-    // this.body.emitter.on("afterDrawing", (ctx) => {this._drawForces(ctx);});
+    //this.body.emitter.on("afterDrawing", (ctx) => {this._drawForces(ctx);});
```
[the remark this answers is not in the document]
Comment (wimrijnders) [reply lib/network/modules/PhysicsEngine.js]: No reason whatsoever. My fingers were probably thinking for me. I actually prefer the space.
Ah yes, I enabled that to test the call, then I added the //'s back again.

File: lib/network/modules/components/physics/BarnesHutSolver.js (hunk 2 above)
Comment (wimrijnders) [reply lib/network/modules/components/physics/BarnesHutSolver.js]: OK. Well, there is a force contribution to be gotten, only it's not returned but added to param `parentBranch`. I can't think of any better names right now.

**Comment by wimrijnders** [comment]
It's actually quite instructive to enable the debug routine for `BarnesHutSolver`. It gives a visual indication of what is actually happening:

download

download 1

Apparently, the optimization here is to subdivide the area and deal with the forces in each area separately. I can sort of intuit that this works.

**Pull request merged**